### PR TITLE
GI-75 Only Approved ideas can have comments

### PR DESCRIPTION
**Pull request by davidkelliott**
We do not want comments to be added to an idea until the idea has been approved.
From this point onwards they are publicly viewable and can be commented on.

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -13,7 +13,11 @@ def index
   def show; end
 
   def new
-    @comment = @idea.comments.build
+    if @idea.approved?
```
Comment (heidar) [inline app/controllers/comments_controller.rb]: It might be better to make this a guard clause to avoid a lengthy if..else like we have in the `create` method?

```
redirect_to(@idea, notice: 'Comments can only be added to approved ideas.') unless @idea.approved?
return if performed?

@comment = @idea.comments.build
```

This is probably more applicable in the `create` method, but applies here too.

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -32,12 +36,16 @@ def destroy
   end
 
   def create
-    @comment = @idea.comments.build(comment_params)
-    @comment.user = current_user
-    if @comment.save
-      redirect_to idea_comment_path(@comment.idea, @comment), notice: 'Comment created'
+    if @idea.approved?
```
Comment (heidar) [inline app/controllers/comments_controller.rb]: Here you could use the pattern I demonstrated on the `new` method to avoid the lengthy if..else and getting rid of the nesting. :)

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -32,12 +36,16 @@ def destroy
...
+      @comment = @idea.comments.build(comment_params)
+      @comment.user = current_user
+      if @comment.save
```
Comment (heidar) [inline app/controllers/comments_controller.rb]: You can also apply it here

File: app/models/idea.rb [hunk 4]
```diff
@@ -17,6 +17,12 @@ def submitted?
     submission_date.present?
   end
 
+  def approved?
+    return false if status == 'awaiting_approval' || status.nil?
```
Comment (heidar) [inline app/models/idea.rb]: How about we default newly created ideas to `awaiting_approval`? That way we can get rid of this method and call the `#awaiting_approval?` method provided by enums.

It's also worth noting that this method overrides the `#approved?` method created by the enum helpers. Could cause confusion in the future.

File: db/seeds.rb [hunk 5]
```diff
@@ -2,20 +2,23 @@
 
 require 'faker'
 
+puts 'Creating admin user admin@justice.gov.uk/change_me'
```
Comment (heidar) [inline db/seeds.rb]: Are these prinouts/changes related to this PR at all? might be worth doing a separate PR for them if not? :D Might be worth using the opportunity to switch the seeds to using factories.

File: db/seeds.rb [hunk 6]
```diff
@@ -63,11 +70,13 @@
...
 
-100.times do
-  idea = Idea.offset(rand(Idea.count)).first
-  user = User.offset(rand(User.count)).first
-  idea.comments.create!(
-    body: Faker::ChuckNorris.fact,
-    user: user
-  )
+puts 'Creating comment for each approved idea'
+Idea.find_each do |idea|
```
Comment (heidar) [inline db/seeds.rb]: Nice use of `.find_each`!

File: db/seeds.rb [hunk 7]
```diff
@@ -63,11 +70,13 @@
...
+  if idea.approved?
```
Comment (heidar) [inline db/seeds.rb]: It might be better to use factories and have one for `:idea, :with_comments` that takes a transient value for the number of comments, then call that a few times explicitly and remove the logic from the seeds.

Can leave this in for this PR but maybe worth doing another one for switching the seeds to use factories instead? :)

File: spec/system/add_comment_spec.rb [hunk 8]
```diff
@@ -57,9 +69,9 @@
 
     describe 'viewing comments' do
       it 'shows comments under the idea' do
-        idea.comments.create!(body: 'Comment 1', user: default_user)
-        idea.comments.create!(body: 'Comment 2', user: default_user)
-        visit idea_path(idea)
+        approved_idea.comments.create!(body: 'Comment 1', user: default_user)
```
Comment (heidar) [inline spec/system/add_comment_spec.rb]: Probably worth using factories for this? :)

File: app/controllers/comments_controller.rb (hunk 1 above)
Comment (davidkelliott) [reply app/controllers/comments_controller.rb]: Done 👍

File: app/controllers/comments_controller.rb (hunk 2 above)
Comment (davidkelliott) [reply app/controllers/comments_controller.rb]: Done 👍

File: app/controllers/comments_controller.rb (hunk 3 above)
Comment (davidkelliott) [reply app/controllers/comments_controller.rb]: Done 👍

File: app/models/idea.rb (hunk 4 above)
Comment (davidkelliott) [reply app/models/idea.rb]: Renamed method to approved_by_admin? to avoid overriding the enum method.  I'll change the statement to 'awaiting_approval' || 'draft' once that change is complete.

File: db/seeds.rb (hunk 5 above)
Comment (davidkelliott) [reply db/seeds.rb]: No not related, removed and will add in a separate PR

File: db/seeds.rb (hunk 7 above)
Comment (davidkelliott) [reply db/seeds.rb]: Will leave for now and create a new PR for changing seeds to factories :)

File: spec/system/add_comment_spec.rb (hunk 8 above)
Comment (davidkelliott) [reply spec/system/add_comment_spec.rb]: I'll create a new PR and card :)
